src/main/java/com/google/devtools/build/lib/remote/blobstore/CombinedDiskHttpBlobStore.java

```diff
@@ -14,136 +14,132 @@
-                  // case of machine
-                  // crashes (the OS may reorder the writes and the rename).
-                  temp.renameTo(target);
-
-                  SettableFuture<Boolean> f = SettableFuture.create();
-                  try (InputStream in = target.getInputStream()) {
-                    ByteStreams.copy(in, out);
-                    f.set(true);
-                  } catch (IOException e) {
-                    f.setException(e);
-                  }
-                  return f;
-                }
-              },
-              MoreExecutors.directExecutor());
-      chained.addListener(
-          () -> {
-            try {
-              tempOut.close();
-            } catch (IOException e) {
-              // not sure what to do here, we either are here because of another exception being
-              // thrown,
-              // or we have successfully used the file we are trying (and failing) to close
-              logger.log(Level.WARNING, "Failed to close temporary file on get", e);
-            }
-          },
-          MoreExecutors.directExecutor());
-      return chained;
+  public void put(String key, long length, InputStream in)
+      throws IOException, InterruptedException {
+    diskCache.put(key, length, in);
+    try (InputStream inFile = diskCache.toPath(key, /* actionResult= */ false).getInputStream()) {
+      remoteCache.put(key, length, inFile);
     }
   }
 
   @Override
-  public boolean getActionResult(String key, OutputStream out)
-      throws IOException, InterruptedException {
-    if (super.getActionResult(key, out)) {
-      return true;
-    }
+  public void putActionResult(String key, byte[] in) throws IOException, InterruptedException {
+    diskCache.putActionResult(key, in);
+    remoteCache.putActionResult(key, in);
+  }
 
-    try (ByteArrayOutputStream tmpOut = new ByteArrayOutputStream()) {
-      if (bsHttp.getActionResult(key, tmpOut)) {
-        byte[] tmp = tmpOut.toByteArray();
-        super.putActionResult(key, tmp);
-        ByteStreams.copy(new ByteArrayInputStream(tmp), out);
-        return true;
-      }
-    }
+  @Override
+  public void close() {
+    diskCache.close();
+    remoteCache.close();
+  }
 
-    return false;
+  @Override
+  public ListenableFuture<Boolean> get(String key, OutputStream out) {
+    return get(key, out, /* actionResult= */ false);
   }
 
   @Override
-  public void put(String key, long length, InputStream in)
-      throws IOException, InterruptedException {
-    super.put(key, length, in);
-    try (InputStream inFile = toPath(key).getInputStream()) {
-      bsHttp.put(key, length, inFile);
+  public ListenableFuture<Boolean> getActionResult(String key, OutputStream out) {
+    return get(key, out, /* actionResult= */ true);
+  }
+
+  private ListenableFuture<Boolean> get(String key, OutputStream out, boolean actionResult) {
+    boolean foundOnDisk = diskCache.toPath(key, actionResult).exists();
```

Comment:
why not call contains here?

Reply:
because it's not aware of the 'actionResult' flag, and I don't want to change an interface at this very moment

Reply:
I think it would be cleaner to update the interface.